### PR TITLE
Multishard

### PR DESCRIPTION
**Pull request by catalinnnn**

### CODE REVIEW

File: common/modules/info_sync/src/info_sync.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#![no_std]
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+const ACCEPT_INFO_ENDPOINT_NAME: &[u8] = b"acceptInformation";
+
+// #[endpoint(takeActionOnInformationReceive)]
```
Comment (bogdan-rosianu) [inline common/modules/info_sync/src/info_sync.rs:8]: remove commented code

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: dex/elrond_dex_pair/src/sharer.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    + token_supply::TokenSupplyModule
+    + token_send::TokenSendModule
+    + liquidity_pool::LiquidityPoolModule
+    + multitransfer::MultiTransferModule
+{
+    #[endpoint(shareInformation)]
+    fn share_information(&self) -> SCResult<()> {
+        let block = self.blockchain().get_block_nonce();
+
+        if block > self.last_info_share_block().get() + self.info_share_min_blocks().get() {
```
Comment (sasurobert) [inline dex/elrond_dex_pair/src/sharer.rs:68]: we must share information or let information share if the difference between local and virtual view is too much

**Review by sasurobert — COMMENTED** [review]
We have to use the price from the oracle modul - using current price is wrong can be attacked

File: dex/elrond_dex_pair/src/config.rs [hunk 3]
```diff
@@ -64,6 +64,13 @@ pub trait ConfigModule: token_send::TokenSendModule {
         self.try_set_fee_percents(total_fee_percent, special_fee_percent)
     }
 
+    #[endpoint(setInfoShareMinEpochs)]
+    fn set_info_share_min_epochs(&self, min_epochs: u64) -> SCResult<()> {
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply dex/elrond_dex_pair/src/config.rs:68]: already fixed in farm pr

File: .github/workflows/actions.yml [hunk 4]
```diff
@@ -21,7 +21,7 @@ jobs:
           source .github/workflows/env
           pip3 install erdpy
           mkdir $HOME/elrondsdk
-          erdpy config set dependencies.arwentools.tag v1.4.1
+          erdpy config set dependencies.arwentools.tag v1.4.14
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply .github/workflows/actions.yml:24]: will upgrade

File: common/modules/info_sync/src/info_sync.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    #[endpoint(addClone)]
+    fn add_clone(&self, clone_address: Address) -> SCResult<()> {
+        require!(
+            !self.clones().contains(&clone_address),
+            "Adress already added"
+        );
+
+        let my_address = self.blockchain().get_sc_address();
+        let my_shard = self.blockchain().get_shard_of_address(&my_address);
+        let clone_shard = self.blockchain().get_shard_of_address(&clone_address);
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply common/modules/info_sync/src/info_sync.rs:23]: ok

File: common/modules/info_sync/src/info_sync.rs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+        if self.received_info().len() == self.clones().len() {
+            self.take_action();
+        }
+
+        Ok(())
+    }
+
+    fn take_action(&self) {
+        self.send().execute_on_dest_context_raw(
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply common/modules/info_sync/src/info_sync.rs:75]: its a generic module. it does not know how the contract looks. did manage to get it to work just with having know a fn signature and with execute on dest for self

File: dex/elrond_dex_pair/src/lib.rs [hunk 7]
```diff
@@ -317,6 +331,8 @@ pub trait Pair:
         )?;
 
         self.burn_tokens(&token_id, &liquidity);
+        let total_liquidity = &self.liquidity().get() - &liquidity;
+        self.liquidity().set(&total_liquidity);
```
Comment (claudiulataretu) [inline dex/elrond_dex_pair/src/lib.rs:335]: Don't you need to extract the liquidity from the total_virtual_liquidity too?

File: dex/elrond_dex_pair/src/lib.rs [hunk 8]
```diff
@@ -475,14 +495,33 @@ pub trait Pair:
         let old_k = self.calculate_k_for_reserves();
 
         let mut reserve_token_out = self.pair_reserve(&token_out).get();
+        let mut virtual_reserve_token_out = self.pair_virtual_reserve(&token_out).get();
         require!(
             reserve_token_out > amount_out_min,
             "Insufficient reserve for token out"
         );
+        require!(
```
Comment (claudiulataretu) [inline dex/elrond_dex_pair/src/lib.rs:503]: I would move this check above the check with reserve_token_out. From my understanding, virtual_reserve_token_out should be always equal or grater than the reserve_token_out. So if virtual_reserve_token_out < amount_out_min then it should be clear that you don't have enough liquidity to perform this swap in any of the siblings contracts

File: dex/elrond_dex_pair/src/liquidity_pool.rs [hunk 9]
```diff
@@ -205,46 +245,94 @@ pub trait LiquidityPoolModule:
...
+        let first_token_price_virtual_max = first_token_price_virtual
+            * (price_percent_total + price_threshold_percent).into()
+            / price_percent_total.into();
+
+        first_token_price_local > first_token_price_virtual_min
+            && first_token_price_local < first_token_price_virtual_max
+    }
+
+    fn swap_too_big(&self, amount: &Self::BigUint, reserve: &Self::BigUint) -> bool {
+        amount > &(reserve / &10u64.into())
```
Comment (claudiulataretu) [inline dex/elrond_dex_pair/src/liquidity_pool.rs:335]: why 10? Could be any other number here?

File: dex/elrond_dex_pair/src/lib.rs [hunk 10]
```diff
@@ -475,14 +495,33 @@ pub trait Pair:
...
+            "Insufficient virtual reserve for token out"
+        );
 
         let mut reserve_token_in = self.pair_reserve(&token_in).get();
-        let amount_out_optimal =
-            self.get_amount_out(&amount_in, &reserve_token_in, &reserve_token_out);
+        let mut virtual_reserve_token_in = self.pair_reserve(&token_in).get();
+
+        let (reserve_token_in_to_use_for_price, reserve_token_out_to_use_for_price) = if self
+            .local_and_virtual_price_differ_too_much()
```
Comment (claudiulataretu) [inline dex/elrond_dex_pair/src/lib.rs:512]: This price function have the following check
`
        first_token_price_local > first_token_price_virtual_min
            && first_token_price_local < first_token_price_virtual_max`
`
So if the local price is between the boundaries for virtual price threshold it means that you want to do the swap with the virtual reserves? If this is the case, then you might need to negate this function result

Comment (catalinnnn) [reply dex/elrond_dex_pair/src/lib.rs:512]: Need to negate local_and_virtual_price_differ_too_much indeed

File: dex/elrond_dex_pair/src/lib.rs [hunk 11]
```diff
@@ -475,14 +495,33 @@ pub trait Pair:
...
         );
+        require!(
+            virtual_reserve_token_out > amount_out_min,
+            "Insufficient virtual reserve for token out"
+        );
 
         let mut reserve_token_in = self.pair_reserve(&token_in).get();
-        let amount_out_optimal =
-            self.get_amount_out(&amount_in, &reserve_token_in, &reserve_token_out);
+        let mut virtual_reserve_token_in = self.pair_reserve(&token_in).get();
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply dex/elrond_dex_pair/src/lib.rs:509]: Yep

File: dex/elrond_dex_pair/src/liquidity_pool.rs (hunk 9 above)
Comment (catalinnnn) [reply dex/elrond_dex_pair/src/liquidity_pool.rs:335]: Might change, will settle later

File: dex/elrond_dex_pair/interaction/testnet.snippets.sh [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    erdpy --verbose contract call $1 --recall-nonce \
+        --pem=$2 \
+        --gas-limit=300000000 \
+        --proxy=${PROXY} --chain=${CHAIN_ID} \
+        --function=addLiquidity \
+        --arguments $3 $4 $5 $6 \
+        --send || return
+}
+
+setLpTokenIdentifier() {
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply dex/elrond_dex_pair/interaction/testnet.snippets.sh:69]: Will fix

File: dex/elrond_dex_pair/src/lib.rs [hunk 13]
```diff
@@ -233,12 +245,14 @@ pub trait Pair:
...
         };
+
+        self.update_safe_reserve();
         self.emit_add_liquidity_event(
             caller,
             first_token_amount.clone(),
             second_token_amount.clone(),
             lp_token_amount.clone(),
-            self.get_total_lp_token_supply(),
+            total_liquidity,
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply dex/elrond_dex_pair/src/lib.rs:255]: Will settle. Good point

**Pull request closed without merging**